wfe/wfe.go

```diff
@@ -1489,22 +1496,56 @@ func (wfe *WebFrontEndImpl) FinalizeOrder(
 		return
 	}
 
+	// spliting order identifiers per types
+	var orderDNSs []string
+	var orderIPs []net.IP
+	for _, ident := range orderIdentifiers {
+		switch ident.Type {
+		case acme.IdentifierDNS:
+			orderDNSs = append(orderDNSs, ident.Value)
+		case acme.IdentifierIP:
+			orderIPs = append(orderIPs, net.ParseIP(ident.Value))
+		default:
+			wfe.sendError(acme.UnauthorizedProblem(
+				fmt.Sprintf("Order includes unknown identifier type %s", ident.Type)), response)
+		}
+	}
+	// sort and deduplicate Names
+	orderDNSs = uniqueLowerNames(orderDNSs)
+	orderIPs = uniqueIPs(orderIPs)
```

Comment:
I think the DNS SAN and IP address identifier deduplication for the order should be done in `NewOrder`. Only the CSR needs to be deduplicated at finalization point. The storage should only create orders in normalized form.

Reply:
I try moving them to NewOrder, but it's look like order is not preserved in storage, failing at CSR test.

Reply:
Ok, that makes sense. Thanks for trying!